#159053991 Allow Users to Favourite and  Unfavourite Article

munyili-samuel opened the pull request:
#### What does this PR do
- Allows users to favourite and unfavourite an article

#### Description of tasks done
- added DB models for favouriting article
- Added urls endpoint for favouriting article
- added views to handle favouriting and un favouriting
- made migrations for database model changes
- added unit tests for this functionality

#### How can this PR be manually tested
Prerequisite: Postman
- use git pull to get the source code from ft-favourite-article-159053991 branch
- use `source .env` to export environmental variables and activate virtual env
- run `python manage.py migrate` to apply recent migrations
- run python manage.py runserver` to start  the project
- if not already done, create a new user account and login
- if not already done, create an article and copy the `slug`
- create a `POST` request to the following endpoint `<str:slug>/favorite/` replacing the `<slug>` with the slug you copied above
- you should get the article back. If you try again you should get a message `You have already marked this article as a favourite`
- create a `DELETE` request to the following endpoint `<str:slug>/favorite/` replacing the `<slug>` with the slug you copied above
- you should get the article back. If you try again you should get a message `You have not marked this article as a favourite`
- if you use a slug that is not valid, you should get a message `The article was not found` on both requests

#### Related PT stories
- [#159053991](https://www.pivotaltracker.com/story/show/159053991)

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
ft

munyili-samuel committed:
change endpoints to use post and delete, not PUT

munyili-samuel committed:
change endpoints to use post and delete, not PUT

munyili-samuel committed:
change endpoints to use post and delete, not PUT

munyili-samuel committed:
change endpoints to use post and delete, not PUT

munyili-samuel committed:
[ft #159053991] add unit tests

munyili-samuel committed:
[ft #159053991] add unit tests

munyili-samuel committed:
[ft #159053991] add unit tests

munyili-samuel committed:
[ft #159053991] add unit tests

munyili-samuel committed:
resolve merge conflicts

munyili-samuel committed:
refactor code

munyili-samuel committed:
refactor code

munyili-samuel committed:
refactor code

munyili-samuel committed:
refactor code

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

munyili-samuel committed:
resolve conflicting migrations

zakhusayn committed:
#159570555 Refactor Authentication app to generic views (#20)

* convert to generic views

munyili-samuel committed:
resolve merge conflicts

RuthNjeri reviewed (changes requested):
Hi Sam,
When I favourite an article using a POST request
```
http://127.0.0.1:8000/api/v1/articles/how-to-train-the-dragon-in-a-ditch-7895168a9cbd41b8ac1822cb729766c0/favorite/
```
 I am getting this as the response 
``` 
{
    "article": {}
}
```
and when unfavourite an article, I get this back
```
{
    "message": {}
}
```
Shouldn't I be getting the article back?

RuthNjeri commented:
When I `GET an article` after favoriting it, I get this response
```
    "favorited": [
        2
    ]
```
Is there a reason you have chosen to return the user id of a person who has favorited an article as opposed to their username?

munyili-samuel commented:
I will fix this in the next 30min.

munyili-samuel committed:
change article serializer to include no of times favourited

munyili-samuel committed:
change article serializer to include no of times favourited

munyili-samuel committed:
change article serializer to include no of times favourited

munyili-samuel committed:
refactor-code: optmise imports

munyili-samuel committed:
refactor-code: optmise imports

munyili-samuel committed:
refactor-code: optmise imports

munyili-samuel committed:
resolve merge conflifts

munyili-samuel committed:
update tests to use new serializer

munyili-samuel committed:
update tests to use new serializer

munyili-samuel committed:
update tests to use new serializer

davenjeru reviewed (approved):
The PR has been tested and is okay. It has been stated that the end user does not need to know who has favourited an article, they just need to see their favourited articles.

davenjeru commented:
It is not a big deal, but who else feels weary about the use of American English on the API endpoints 😅😅😅, Like **favorite** instead of **favourite**. 🤷‍♂️

lennykioko commented on authors/apps/articles/views.py:
```diff
@@ -112,3 +108,82 @@ def update(self, request, slug):
+    def post(self, request, slug):
+        """
+        Helps user favourite article
+        If already, return message saying that the user has already favourated
```
Typo: you meant favourited instead of favourated

munyili-samuel committed:
correct typo in docsrings, update serializer docstring

munyili-samuel committed:
correct typo in docsrings, update serializer docstring

munyili-samuel committed:
correct typo in docsrings, update serializer docstring

davenjeru committed:
#159053987 Add functionality for rating articles (#24)

-  add tests to test for the rating functionality
-  add functionality to view an article's ratings and rate an article
- update the README to reflect the added functionality

lennykioko committed:
Update README to reflect current API spec (#29)

* Add instructions for testing the application

hoslack committed:
#159053988 Add functionality for users to be able to comment on articles (#28)

- Add endpoints for creating and deleting threaded comments
- fix slug to remain the same on article edit

RuthNjeri committed:
[ch #159762262] add Hound yml file (#31)

lennykioko committed:
#159053984 refactor social_auth (#27)

* Update social_auth url for uniformity
* Update README to reflect current API spec

zakhusayn committed:
#159053994 Pagination for the articles (#30)

* Add configuration for pagination
* Add test for pagination

munyili-samuel committed:
resolve merge conflicts

munyili-samuel committed:
fix: delete conflicting migrations

munyili-samuel committed:
fix: delete conflicting migrations

munyili-samuel committed:
fix: recreate migrations

munyili-samuel committed:
fix: recreate migrations

dmigwi reviewed (changes requested):
Please fix all the pep8 violations on your PR.

munyili-samuel committed:
ch: reformat modules to comply with PEP8

munyili-samuel committed:
ch: optmize imports, reformat file

munyili-samuel committed:
ch: fix PEP8 violations

munyili-samuel committed:
reformat code to adhere to PEP8

munyili-samuel committed:
reformat code to adhere to PEP8

munyili-samuel committed:
reformat code to adhere to PEP8

munyili-samuel committed:
reformat code to adhere to PEP8

munyili-samuel committed:
fix: change permission class for articles views to fix tests

dmigwi commented:
How did you end up making 87 commits?
